src/content/docs/en/guides/view-transitions.mdx

```diff
@@ -465,20 +465,79 @@ If you have code that sets up global state, this state will need to take into ac
 
 Module scripts are only ever executed once because the browser keeps track of which modules are already loaded. For these scripts, you do not need to worry about re-execution.
 
-### `astro:page-load`
+## Lifecycle events
 
-An event that fires at the end of page navigation, after the new page is visible to the user and blocking styles and scripts are loaded. You can listen to this event on the `document`.
+The `<ViewTransition />` router fires a number events on the `document` during navigation. These events provide hooks into the lifecycle of navigation, allowing to do things like showing indicators that a new page is loading, override default behavior, and restore state as navigation is completing.
 
-The `<ViewTransitions />` component fires this event both on initial page navigation (MPA) and any subsequent navigation, either forwards or backwards.
+### `astro:before-preparation`
 
-You can use this event to run code on every page navigation, or only once ever:
+<Since v="3.6.0" />
 
-```astro "{ once: true }"
-<script>
-  document.addEventListener('astro:page-load', () => {
-    // This only runs once.
-    setupStuff();
-  }, { once: true });
+An event that fires at the beginning of the preparation phase, after navigation has started (for example, after the user has clicked a link), but before content is loaded.
+
+This event is useful for a few use-cases:
+
+- To do something before loading has started, such as showing an a loading spinner.
+- To alter loading, such as loading content you've defined in a template rather than from the external URL.
+- To change the `direction` of the navigation (which is usually either `forward` or `backward`) in order to do custom animation.
+
+Here's an example of using the `astro:before-preparation` event to load a spinner before the content is loaded.
+
+```js
+<script is:inline>
+  document.addEventListener('astro:before-preparation', ev => {
+    const originalLoader = ev.loader;
+    ev.loader = async function() {
+      const { startSpinner } = await import('./spinner.js');
+      const stop = startSpinner();
+      await originalLoader();
+      stop();
+    };
+  });
+</script>
+```
+
```

Comment:
I like this example where the spinner is set and removed with the loader callback.    But if there is another wrapper around it that needs some time to do e.g. further prefetching based on the loaded data, then the spinner is gone too soon. It would be much better to remove the spinner in astro:after-preparation, because all loaders/prefetchers/... are through and the screenshot has not yet been taken.

Reply:
Is this easy to show in a code sample? Do we just show **starting** the spinner here? I will ask you or Matthew to edit the sample to your liking!

Reply:
@martrapp Given @matthewp 's new example below using both `before-` and `after-` preparation, should we remove this example altogether, and let the combined example below serve for both?